napari/layers/labels/labels.py

```diff
@@ -682,46 +734,64 @@ def paint(self, coord, new_label, refresh=True):
             self._save_history()
+        brush_size_dims = [self.brush_size] * self.ndim
+        if not self.n_dimensional and self.ndim > 2:
+            for i in self.dims.not_displayed:
+                brush_size_dims[i] = 1
 
-        if self.n_dimensional or self.ndim == 2:
+        if self.brush_shape == "square":
             slice_coord = tuple(
-                [
-                    slice(
-                        np.round(
-                            np.clip(c - self.brush_size / 2 + 0.5, 0, s)
-                        ).astype(int),
-                        np.round(
-                            np.clip(c + self.brush_size / 2 + 0.5, 0, s)
-                        ).astype(int),
-                        1,
-                    )
-                    for c, s in zip(coord, self.shape)
-                ]
-            )
-        else:
-            slice_coord = [0] * self.ndim
-            for i in self.dims.displayed:
-                slice_coord[i] = slice(
-                    np.round(
-                        np.clip(
-                            coord[i] - self.brush_size / 2 + 0.5,
-                            0,
-                            self.shape[i],
-                        )
-                    ).astype(int),
-                    np.round(
-                        np.clip(
-                            coord[i] + self.brush_size / 2 + 0.5,
-                            0,
-                            self.shape[i],
-                        )
-                    ).astype(int),
+                slice(
+                    np.round(np.clip(c - brush_size / 2 + 0.5, 0, s)).astype(
+                        int
+                    ),
+                    np.round(np.clip(c + brush_size / 2 + 0.5, 0, s)).astype(
+                        int
+                    ),
                     1,
                 )
-            for i in self.dims.not_displayed:
-                slice_coord[i] = np.round(coord[i]).astype(int)
+                for c, s, brush_size in zip(coord, self.shape, brush_size_dims)
+            )
+        elif self.brush_shape == "circle":
+            slice_coord = [int(np.round(c)) for c in coord]
+            shape = self.shape
+            if not self.n_dimensional and self.ndim > 2:
+                coord = [coord[i] for i in self.dims.displayed]
+                shape = [shape[i] for i in self.dims.displayed]
+
+            sphere_dims = len(coord)
+            # Ensure circle doesn't have spurious point
+            # on edge by keeping radius as ##.5
+            radius = np.floor(self.brush_size / 2) + 0.5
+            mask_indices = self.sphere_indices(radius, sphere_dims)
+
+            mask_indices = mask_indices + np.round(np.array(coord)).astype(int)
+
+            # discard candidate coordinates that are out of bounds
+            discard_coords = np.logical_and(
+                ~np.any(mask_indices < 0, axis=1),
+                ~np.any(mask_indices > np.array(shape), axis=1),
```

Comment:
I was getting a bug on the right/bottom edge of the array when painting in `examples/add_labels.py`. That's because this should be:

```python
np.all(mask_indices < np.array(shape), axis=1),
```

(or, less intrusive:

```python
~np.any(mask_indices >= np.array(shape), axis=1),
```
, note the `>=` rather than `>`)

Additionally, these are not the discard coords, but rather the kept coords!

Reply:
Incidentally, worth writing a test for this, then fixing!